--- a/src/couch/priv/couch_cfile/couch_cfile.c
+++ b/src/couch/priv/couch_cfile/couch_cfile.c
@@ -14,7 +14,7 @@

 #include "erl_nif.h"

-#ifndef _WIN32
+#if !defined(_WIN32) && defined(__LP64__)
 #define COUCH_CFILE_SUPPORTED 1
 #endif

--- a/src/couch/src/couch_file.erl
+++ b/src/couch/src/couch_file.erl
@@ -985,14 +985,20 @@ maybe_close_cfile_handle(Fd) ->

 dup(#file{fd = Fd} = File) ->
     case config:get_boolean("couchdb", "use_cfile", ?USE_CFILE_DEFAULT) of
         true ->
-            case couch_cfile:dup(Fd) of
-                {ok, CFd} ->
-                    ok = file:close(Fd),
-                    CFile = File#file{fd = CFd},
-                    put(couch_file_fd, {CFd, CFile#file.filepath}),
-                    put(?CFILE_HANDLE, CFile#file{eof = 1 bsl 60}),
-                    CFile;
-                {error, _Error} ->
+            try
+                case couch_cfile:dup(Fd) of
+                    {ok, CFd} ->
+                        ok = file:close(Fd),
+                        CFile = File#file{fd = CFd},
+                        put(couch_file_fd, {CFd, CFile#file.filepath}),
+                        put(?CFILE_HANDLE, CFile#file{eof = 1 bsl 60}),
+                        CFile;
+                    {error, _Error} ->
+                        File
+                end
+            catch
+                error:undef ->
+                    % couch_cfile NIF not available (32-bit ARM systems)
                     File
+            end;
         false ->
             File
